charon/src/bin/charon-driver/translate/translate_types.rs

```diff
@@ -779,7 +781,101 @@ impl<'tcx, 'ctx> ItemTransCtx<'tcx, 'ctx> {
+        let tcx = self.t_ctx.tcx;
+        let rdefid = self.def_id.as_rust_def_id().unwrap();
+        let ty_env = hax::State::new_from_state_and_id(&self.t_ctx.hax_state, rdefid).typing_env();
+        let ty = tcx.type_of(rdefid).skip_binder();
+        let pseudo_input = ty_env.as_query_input(ty);
+
+        match tcx.layout_of(pseudo_input) {
+            Ok(ty_layout) => {
+                let layout = ty_layout.layout;
+                let (size, align) = if layout.is_sized() {
+                    (
+                        Some(layout.size().bytes()),
+                        Some(layout.align().abi.bytes()),
+                    )
+                } else {
+                    (None, None)
+                };
+                let mut variant_layouts = Vector::new();
+                let discriminant_offset = match layout.variants() {
+                    // If the discriminant is in the niche, it isn't part of
+                    // the actual layout.
+                    r_abi::Variants::Multiple {
+                        tag_field,
+                        tag_encoding: r_abi::TagEncoding::Direct,
+                        variants,
+                        ..
+                    } => {
+                        for variant in variants.iter() {
+                            let index = if let r_abi::Variants::Single { index } = variant.variants
+                            {
+                                index
+                            } else {
+                                // The variants have their VariantIndex as the index in `Single`.
+                                unreachable!()
+                            };
+                            variant_layouts.insert(
+                                VariantId::from_usize(index.as_usize()),
+                                translate_variant(variant),
+                            );
+                        }
+
+                        if let r_abi::FieldsShape::Arbitrary { offsets, .. } = layout.fields() {
+                            offsets
+                                .get(r_abi::FieldIdx::from_usize(*tag_field))
+                                .map(|s| r_abi::Size::bytes(*s))
+                        } else {
+                            // The tag_field is the index into the fields vector,
+                            // thus, the fields shape must be arbitrary.
+                            unreachable!()
+                        }
+                    }
+                    _ => None,
+                };
+
+                Layout {
+                    size,
+                    align,
+                    discriminant_offset,
+                    variant_layouts,
+                }
+            }
+            Err(_) => {
+                // If there is no sensible layout, we translate it to the empty layout.
+                // There doesn't seem to be any way to handle the errors in a better way.
+                Layout::default()
```

Comment:
Yeah, as I've said elsewwhere this isn't good: we should make sure to have the right number of variants and in each the right number of fields. Alternatively we could just store `Option<Layout>`, that's also ok.